### ISSUE TITLE
set `CXX_STANDARD=17` to ensure C++17 compatibility

### ISSUE THREAD

**Issue by Dominik-K** [EVerest/EVerest-archived#114]
It's useful for multiple purposes:

1. Fixes build-issues for older compilers:
    Some users use Ubuntu 20.04 where only `g++-10` is available and therefore C++17 isn't default, yet ([since g++-11, it's the default](https://gcc.gnu.org/projects/cxx-status.html#cxx17)). See [topic in mailing-list](https://lists.lfenergy.org/g/everest/topic/102236317).
2. Enforces to not accidentally use newer C++-features without offering a backwards-compatibility. A good example for it would be using `std::format` resp. the [`fmt`-library](https://github.com/fmtlib/fmt).

On the other hand, we should offer an easy way to disable this option as some still need support for older C++-versions. See
- https://github.com/EVerest/libocpp/pull/177

## Tasks
- [ ] `libevse-security`
- [ ] `libocpp`
- [ ] ...

**Comment by hikinggrass** [comment]
I think that's something that the libraries actually using c++17 features should set for the targets that need c++17, if that's missing it's a bug that should be fixed in the respective libraries. However I think it's useful to set this in our CI so we catch accidental usage of newer c++ features

**Comment by Dominik-K** [comment]
Yes, that's what I mean. I think it's correct to set this as default and offer an `option(EVEREST_CXX_OLD_VERSION_SUPPORT)` in `CMakeLists.txt` to get backwards-compatibility support (which we should check for in CI, sure).

Another possibility would be to set this in the `CmakePresets.json`, but given the consideration in this [good Stackoverflow answer](https://stackoverflow.com/a/75153894/1168315), the approach above looks better.

**Comment by hikinggrass** [comment]
> Yes, that's what I mean. I think it's correct to set this as default and offer an `option(EVEREST_CXX_OLD_VERSION_SUPPORT)` in `CMakeLists.txt` to get backwards-compatibility support (which we should check for in CI, sure).

Thankfully we do not _really_ have to support anything earlier than C++17 at the moment, the std::filesystem / boost::filesystem thing is actually a special case where some versions of GCC support C++17 but only the language features, not all standard library features

> Another possibility would be to set this in the `CmakePresets.json`, but given the consideration in this [good Stackoverflow answer](https://stackoverflow.com/a/75153894/1168315), the approach above looks better.

CMakePresets.json isn't really an option in your example:

1. CMakePresets.json [needs CMake>=3.19](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html)
2. we support CMake>=3.14
3. Ubuntu 20.04 only [ships with cmake 3.16](https://launchpad.net/ubuntu/focal/+source/cmake)

**Comment by a-w50** [comment]
@Dominik-K What do you want to make optional here?  If a library needs to be compiled with c++17, it should set
`target_compile_features(lib_target_name PRIVATE cxx_std_17)`
if it exposes c++17 features in its headers, it needs to set
`target_compile_features(lib_target_name PUBLIC cxx_std_17)`
so that everyone using this library (at least with cmake) will automatically compile with c++17
In case the library doesn't set these things, the cmake files of the library should be fixed.
Setting `CXX_STANDARD=17` on the other hand will force everything inside the cmake build to be build with c++17, which is usually not, what we want.
